### ISSUE TITLE
[Bug]: Linux PWA doesn't retain its PWA icon after loading, and reverts back the Floorp icon, and gets grouped with it.

### ISSUE THREAD

**Issue by 32bitx64bit** [Floorp-Projects/Floorp#1102]
### Pre-Submission Checklist

- [X] I have checked the existing/closed issues to ensure this issue has not already been reported.
- [X] I am using the latest supported version of the browser.
- [X] I have tried disabling all browser extensions/add-ons to rule out interference from these.
- [X] I have tried clearing the browser cache and cookies.
- [X] I have tried to create a new profile and the issue still occurs.
- [X] I have provided detailed step-by-step instructions on how to reproduce the issue.
- [X] I have included relevant screenshots or console outputs.
- [X] I have checked this problem is not ocurring in the latest version of Firefox ESR.
- [X] This issue is specific to this browser and does not occur in other browsers.

### Bug Summary

When you open a PWA via the taskbar / any other way (This example im on KDE), it will retain its icon until the webpage is fully loaded, then it reverts back to the Floorp icon, and gets grouped with it again.

Here's a video of it happening to better explain 

https://github.com/Floorp-Projects/Floorp/assets/86190528/79fad40a-ffeb-4694-942f-6a07b202d406

### Reproduction Steps

Install any PWA on Linux KDE.

Open PWA 

Done.

### Expected Behavior

Expected behavior should be like on windows where it retains its identity.

### Actual Behavior

https://github.com/Floorp-Projects/Floorp/assets/86190528/66b4f5f2-7ee3-4c74-9258-3389cd732e11

### Screenshots

https://github.com/Floorp-Projects/Floorp/assets/86190528/8b21f027-0ada-4987-bfe8-422e856c52ae

### Browser Version

11.12.0

### Operating System

Arch Linux KDE Plasma 6.0.3

**Comment by Sasikuttan2163** [comment]
Unfortunately that's a limitation of how the PWA was implemented. For this release we focused on creating a desktop entry so that the app gets listed on launchers. I'm currently looking into it.

**Comment by 32bitx64bit** [comment]
> Unfortunately that's a limitation of how the PWA was implemented. For this release we focused only on creating a desktop entry so that the app gets listed on launchers. I'm currently looking into it.

Ah. Hopefully this gets resolved!
But it is nice to see linux PWA's being worked on.

**Comment by b-risk** [comment]
Apologizes for (semi) unrelated question, but how would I go about adding a shortcut like this to a web app in Linux Mint? I've created web apps in Floorp I didn't see an option to add a shortcut for it in Floorp settings.

**Comment by Sasikuttan2163** [comment]
Can you be more specific as to what you mean by create a shortcut? Is it that the app doesn't appear in the all apps list or you can't create a shortcut to the app in the desktop? (Sorry for the late reply as well, I have exams going on)

**Comment by b-risk** [comment]
> Can you be more specific as to what you mean by create a shortcut? Is it that the app doesn't appear in the all apps list or you can't create a shortcut to the app in the desktop? (Sorry for the late reply, I have exams going on)

Yes I should have clarified, I meant like how there isn't shortcuts on Linux but they use launchers instead that appear in the apps list, I don't see an option anywhere to create one of these for my Floorp web apps on my distro, also I should mention I'm using the cinnamon desktop version of Linux Mint.

**Comment by RimPYExposing** [comment]
> > Can you be more specific as to what you mean by create a shortcut? Is it that the app doesn't appear in the all apps list or you can't create a shortcut to the app in the desktop? (Sorry for the late reply, I have exams going on)
> 
> Yes I should have clarified, I meant like how there isn't shortcuts on Linux but they use launchers instead that appear in the apps list, I don't see an option anywhere to create one of these for my Floorp web apps on my distro, also I should mention I'm using the cinnamon desktop version of Linux Mint.

As stated on the PWA, this is experimental, if theirs a bug unrelated to this, create a new issue instead of clogging this one up with unrelated questions. As this currently works fine on KDE plasma 6 (wayland)

**Comment by Jeremi360** [comment]
@Sasikuttan2163 What @b-risk means is that linux mint has it own app called [Webapp-Manger](https://github.com/linuxmint/webapp-manager/tree/master) that allows to crate desktop web app using any of many supported browsers, but floorp is not supported, but it is not a real PWA - they hack normal way of opening normal websites.

I some how their way of doing this dosen't have such problems.
Also I like that floorp in current version allow me to use even discord as web app,
even though it is not a PWA

**Comment by b-risk** [comment]
> @Sasikuttan2163 What @b-risk means is that linux mint has it own app called [Webapp-Manger](https://github.com/linuxmint/webapp-manager/tree/master) that allows to crate desktop web app using any of many supported browsers, but floorp is not supported, but it is not a real PWA - they hack normal way of opening normal websites.
> 
> I some how their way of doing this dosen't have such problems. Also I like that floorp in current version allow me to use even discord as web app, even though it is not a PWA

Yes I know Floorp has a different implementation and that's why I didn't mention that app, I'm just confused on how to create one of these app launchers for Floorp, like how on KDE how do you create them? To be clear, I have several web apps already in Floorp created, they just don't show in the apps list, that's the reason I didn't post this as an issue because I was confused if it was intended behavior or if it's actually a distro specific issue and it's supposed to add automatically to my apps list. Thanks and apologizes again.

**Comment by Sasikuttan2163** [comment]
> Yes I know Floorp has a different implementation and that's why I didn't mention that app, I'm just confused on how to create one of these app launchers that show in the apps list for Floorp web apps, like how on KDE how do you create them? To be clear, I have several web apps already in Floorp created, they just don't show in the apps list, that's the reason I didn't post this as an issue because I was confused if it was intended behavior or if it's actually a distro specific issue and it's supposed to add automatically to my apps list. Thanks and apologies again.

Oh that's weird? Are you saying it doesn't show up in the launcher at all? I just installed Cinnamon myself and found my Floorp web apps in the "Others" sub section within the launcher.

**Comment by b-risk** [comment]
> > Yes I know Floorp has a different implementation and that's why I didn't mention that app, I'm just confused on how to create one of these app launchers that show in the apps list for Floorp web apps, like how on KDE how do you create them? To be clear, I have several web apps already in Floorp created, they just don't show in the apps list, that's the reason I didn't post this as an issue because I was confused if it was intended behavior or if it's actually a distro specific issue and it's supposed to add automatically to my apps list. Thanks and apologies again.
> 
> Oh that's weird. Are you saying it doesn't show up in the app list at all? I just installed Cinnamon myself and found my Floorp web apps in the "Others" sub section within the launcher.

Ah, I wasn't sure so it's just automatic, I just tried it again and now it shows in the apps list, I was sure I had tried several times... maybe my flatpak had updated without me noticing? Anyways thanks for the help!

**Comment by Sasikuttan2163** [comment]
Glad it got fixed! 😄

**Comment by luvis** [comment]
this icon "issue" is actually the only thing i feel missing in floorp for it to replace my current daily driver. I really hope this can be fixed as I would finally be able to leave chromium =) Is there any update to this?

**Comment by Ramen-LadyHKG** [comment]
I think the only way to fix it, is to create a new desktop file automatically when PWA is installed.
At least that's what I found on Chromium browsers (Brave, Vivaldi)

![Image](https://github.com/user-attachments/assets/3ec4b5cc-28a2-4adb-a97c-adae669f8ed4)

**Comment by acorazza** [comment]
I have the same issue, and even trying to set a custom icon in KDE settings doesn't work, the Floorp icon remains.

**Comment by TheChilledBuffalo** [comment]
> I think the only way to fix it, is to create a new desktop file automatically when PWA is installed. At least that's what I found on Chromium browsers (Brave, Vivaldi)

@Ramen-LadyHKG  Actually, Floorp creates new .desktop files too...

**Comment by CiRiX-** [comment]
@surapunoyousei  Is this something that will be worked on?
This is the reason as to why i need to use Edge for my Web-Apps.  😢

**Comment by teohhanhui** [comment]
Investigating a little...

If you run Floorp with e.g.

```shell
one.ablaze.floorp --name='SomeName'
```

You can verify that it has the `WM_CLASS` (or its Wayland equivalent `app_id`) of `SomeName`, e.g. on GNOME by pressing `Alt`+`F2` and running `lg` and look in the Windows tab.

However, it seems this only works for the first window opened for the same profile.

And because Floorp uses the same profile for running PWAs, trying to set `--name` in the `Exec=` line of the PWA's desktop file doesn't do anything.

Maybe there's a way for Floorp to set the `WM_CLASS` / `app_id` when launching a PWA?

This seems like where it's actually set: https://github.com/mozilla-firefox/firefox/blob/4e0efb194887d98fb4a7c43eb495e4b6ecd47795/widget/gtk/nsWindow.cpp#L6523-L6628

**Comment by Sasikuttan2163** [comment]
@CiRiX- unfortunately afaik it is a Firefox limitation. @TheChilledBuffalo and I did look into it back when we were implementing it for Floorp 11. Edge is based on Chromium.

**Comment by teohhanhui** [comment]
@Sasikuttan2163 There's no "Firefox limitation", as Floorp is a Firefox fork. It can be implemented for sure, just need to find the right place to do it.

**Comment by Sasikuttan2163** [comment]
It was not my intention to say that it is impossible to implement. The [PR](https://github.com/Floorp-Projects/Floorp-core/pull/63) which ported PWA support to Linux added desktop entry support so that users could launch web apps from their application launchers. As you can see in the PR we only modified the JS code of the browser to make it happen.

As for the app icons problem, we tried a few things with browser-level code but they did not work. Any solution to this problem would involve a significant rewrite of the underlying CPP architecture of the browser. Modifying the architecture will definitely fix this problem, however a rewrite of this scale would definitely be harder to maintain and would likely need atleast one dedicated Linux user among the devs to test and patch out any bugs before every update. 

I appreciate the effort you put in to find the probable cause and solution to the issue but you should also understand that supporting a feature like this is a very big commitment for the devs. This isn't a fix-and-forget type of situation. I hope you understand.

**Comment by teohhanhui** [comment]
Not stale.

**Comment by Ryosuke-Asano** [comment]
I did it https://blog.floorp.app/ja/release/12.7.0/

**Comment by paulius-valiunas** [comment]
@Ryosuke-Asano I'm not sure what you fixed, maybe there's something I missed... Microsoft Teams still uses the Floorp icon in my taskbar. The desktop entry has the correct icon, but once I launch it, it reverts to Floorp.

**Comment by Devilicate** [comment]
Yeah issue also remains for me, at least with YouTube Music.
